pkg/apis/resource/types.go

```diff
@@ -0,0 +1,466 @@
+	// Claim.
+	Spec ClaimSpec
+
+	// Status describes whether the resource is available and with which
+	// attributes.
+	Status ClaimStatus
+}
+
+// ClaimSpec defines how a resource is to be allocated.
+type ClaimSpec struct {
+	// ClassName references the driver and additional parameters
+	// via the name of a Class that was created as part of the
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists.
+	ClassName string
+
+	// ParametersRef references a separate object with arbitrary parameters
+	// that will be used by the driver when allocating a resource for the
+	// claim.
+	//
+	// The object must be in the same namespace as the Claim.
+	ParametersRef *ClaimParametersReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. "WaitForFirstConsumer" is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a Claim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod
+// (AllocationModeWaitForFirstConsumer). Other modes might get added in the
+// future.
+type AllocationMode string
+
+const (
+	// When a Claim has AllocationModeWaitForFirstConsumer, allocation is
+	// delayed until a Pod gets scheduled that needs the Claim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeWaitForFirstConsumer AllocationMode = "WaitForFirstConsumer"
+
+	// When a Claim has AllocationModeImmediate, allocation starts
+	// as soon as the Claim gets created. This is done without
+	// considering the needs of Pods that will use the Claim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ClaimStatus struct {
+	// DriverName is a copy of the driver name from the Class at
+	// the time when allocation started. It's necessary to enable usage of
+	// the claim by kubelet in case that the Class got removed in
+	// the meantime. It also helps the resource driver to determine whether
+	// it needs to handle a claim that got marked for deletion.
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
+	// allocated.
+	Allocation *AllocationResult
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. Usually those are Pods, but other objects are also
+	// possible as long as they exist as resources in the apiserver. The
+	// Claim controller in the kube-controller-manager checks for
+	// stale entries that reference non-existent objects and removes them.
+	//
+	// A scheduler must add a Pod that it is scheduling. This must be done
```

Comment:
These are sort of impl details - is there not a better place for this?

Reply:
The KEP?

I don't think we have any in-tree place where "correct usage of the API" can be described. The API description comes closest, but not if we treat it as "for users".

Reply:
@sftim opinions on this?  In particular, should API docs include details about the way that fields are supposed to be used by components like drivers and scheduler?

Reply:
> should API docs include details about the way that fields are supposed to be used by components like drivers and scheduler?

I recommend making that a last resort. The kube-controller-manager in particular is written in a way that lets us replace it with, for example, lots of separate components that each handle a different set of API groups. A split isn't likely but - for the API reference - I'm not keen on documenting a particular way of laying out the control plane when Kubernetes' API doesn't actually require that.

We have quite a few precedents for renaming, splitting, or combining controllers. Putting that much detail here feels like an invitation for things to go stale.